### ISSUE TITLE
spack spec -t can report misleading dependency types in case of multiple dependencies

### ISSUE THREAD

**Issue by healther** [spack/spack#8802]
We noticed this internally with one of our more complicated meta packages. But it is straight forward to reproduce by creating 4 packages:
* a: depends on b and c with either link or run dependency
* b: depends on d type build
* c: depends on d type either link or run
* d: no dependencies

### Expected Result

```
spack spec -t a
```
to correctly identify d being a link/run dependency instead of only being a build dependency.

### Actual Result

```
$ spack spec -t a
Input spec
--------------------------------
[    ]  a

Concretized
--------------------------------
[    ]  a@1%clang@10.0.0-apple arch=darwin-highsierra-x86_64 
[bl  ]      ^b@1.2.3%clang@10.0.0-apple arch=darwin-highsierra-x86_64 
[b   ]          ^d@1.2.3%clang@10.0.0-apple arch=darwin-highsierra-x86_64 
[bl  ]      ^c@1.2.3%clang@10.0.0-apple arch=darwin-highsierra-x86_64 
```

I'll try to track this down either later today or this week, but for now we should tread the type identification of dependencies of dependencies from `spack spec -t` as unreliable.

**Comment by healther** [comment]
Okay now my confusion only increases. I basically stepped through the `spec.concretize()` part of the parser. And as far as I can see the `link` dependency of `c` on `d` is simply gone after the concretisation finishes.

I'm confused, because our `view` generation correctly adds `d` to the view if there is a `link` or `run` type dependency on it anywhere in the dag. But it doesn't when there are only `build` dependencies.

I believe I tracked it to `spec.py:l2172` 
```
    def _normalize_helper(self, visited, spec_deps, provider_index, tests):
        """Recursive helper function for _normalize."""
        if self.name in visited:
            return False
```
this `return False` is only the right thing to do if the dependency types agree other wise the dependency types need to be updated, unfortunately I have no idea how to do this (and neither what the semantics of the output should be then)
In this case a correct DAG would be
```
[    ]  a@1%clang@10.0.0-apple arch=darwin-highsierra-x86_64 
[bl  ]      ^b@1.2.3%clang@10.0.0-apple arch=darwin-highsierra-x86_64 
[bl  ]      ^c@1.2.3%clang@10.0.0-apple arch=darwin-highsierra-x86_64 
[bl  ]          ^d@1.2.3%clang@10.0.0-apple arch=darwin-highsierra-x86_64 
```
but this is only true as long there is no `depends_on('xyz', type='build')` and `depends_on('xyz', type='link')`in packages `xx` and `yy` respectively. Then placing the union of dependencies under one would be ?misleading?

As I'm needing help for this pinging people I think might know what to do @scheibelp @tgamblin

**Comment by tgamblin** [comment]
@healther: you're just seeing an artifact of how `spack spec` renders the DAG by default.  `spack spec` is a tree view of a DAG.  It's intended to make the view more manageable.  By default we only show each node in the DAG once, because if we traverse every edge, the output can get very large.  Consider, for example, that `spack spec` output can *already* get large for things that have lots of dependencies.

If you ask for `-t`, which shows dependency types, you'll only see the dependency types *on the first edge traversed* to a node.  If the traversal encounters the node again, we skip it because it's already been printed.

In your example, the dependency graph looks like this:

```
$ spack graph a
o  a
|\
o |  c
| o  b
|/
o  d
```

There are two edges into `d` and you're only seeing the one from `c` to `d` because it's the first one traversed.

If you use `--cover edges`, you'll see the dependency type on the edge from `b` to `d` as well, and you'll see that it is correct (my graph has a `link` and a `run` dependency to show that both work -- yours just has `link`):

```
$ spack spec -t --cover edges a
Input spec
--------------------------------
[    ]  a

Concretized
--------------------------------
[    ]  a@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64
[ lr ]      ^b@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64
[b   ]          ^d@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64
[ lr ]      ^c@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64
[ lr ]          ^d@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64
```

`--cover nodes` is the default and guarantees only that each node is shown once.  `--cover edges` guarantees that each *edge* is displayed once, so it shows both the `b -> d` and `c -> d`.

For completeness, there is also `--cover paths`, which will covers unique paths in the graph.  for the above graph, `--cover paths` output is the same as for `--cover edges`, but now suppose that `b` also depends on `c`:

```
$ spack graph a
o  a
|\
| o  b
|/| 
o |  c
|/
o  d
```

Now, `--cover edges` shows this:

```
 spack spec -t --cover edges a
Input spec
--------------------------------
[    ]  a

Concretized
--------------------------------
[    ]  a@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64 
[ lr ]      ^b@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64 
[ lr ]          ^c@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64 
[ lr ]              ^d@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64 
[b   ]          ^d@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64 
[ lr ]      ^c@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64 
```

It gets to `c` the first time through `a -> b -> c`, but it doesn't descend past `c` the second time, when it is reached through `a -> c`, as it has already covered the `c -> d` edge.  If you do `--cover paths`, it will *always* descend into sub-graphs even if they've already been covered by another path:

```
$ spack spec -t --cover paths a
Input spec
--------------------------------
[    ]  a

Concretized
--------------------------------
[    ]  a@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64 
[ lr ]      ^b@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64 
[ lr ]          ^c@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64 
[ lr ]              ^d@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64 
[b   ]          ^d@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64 
[ lr ]      ^c@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64 
[ lr ]          ^d@1.0%clang@8.1.0-apple arch=darwin-highsierra-x86_64
```

The advantage of that is that every time a node appears, all of its dependencies are shown.  The obvious disadvantage is that the output can get really long for even moderately complex graphs.

Does that help?

**Comment by tgamblin** [comment]
@healther: The view bug you found should probably be a separate issue -- as it might matter in other situations.  Can you add that?

**Comment by healther** [comment]
Mh okay, this seems reasonable, let check what `view` is then actually doing. I.e. whether `s.traverse()` disregards `build`-only dependencies or not.

I'm not sure if the output from `spack spec` as it is right now is the most useful, what (at least) I am interested when I'm going to `spack spec` is: "What are the dependencies of my partial spec?" I.e. I'm less interested whether `a` depends on `d` through `b` or through `c` than I'm interested in in which ways `a` ends up depending on `d`. I.e. Is it necessary for me to keep `d` around if I want to use `a`.

Would it be possible to print something like an "effective" dependency? I.e. everything `a` only depends on through a `build` or `test` dependency can be removed, but everything `a` has a `link` or `run` dependency on needs to be kept around.

**Comment by tgamblin** [comment]
I mean, that's obviously useful.  If you want to add something like that feel free 😄.  Something merged might be better for the default `-t` view than what is currently there.

**Comment by alalazo** [comment]
> Something merged might be better for the default -t view than what is currently there.

+1 to this. The default view for `d` would be more informative saying `[blr ]` (merging incoming edges) than just `[b   ]`.

**Comment by healther** [comment]
Okay so very strange things are happening (take what I'm about to say with a grain of salt because I take my information from a rather large DAG, that I can't easily isolate; will do a minimal view as soon as I get to it):
`s.traverse()` seems to only follow `link` and `run` edges, even though `get_dependencies` calls it without an explicit `deptype` and `traverse_edges` defaults to `deptype='all'`. Nevertheless `libtool` does not appear in the final view even though there are multiple packages that have a `depends_on('libtool', type='build')` in the requested DAG.

I'm currently even uncertain what the intended behaviour should be. Given that we only noted it because someone wanted to do something with `libtool` and hence was missing it (so having it added would have spared us the hiccup). On the other hand `build`-only dependencies should be able to be removed without affecting anything (which should be currently broken, again I need to get to check this).

Question: If `a` has a `build`-only dependency on `b` and I do `spack install a && spack uninstall b` will spack try to uninstall `a`? (If someone knows by heart otherwise I'll check this with the other stuff I need to test)

> I mean, that's obviously useful. If you want to add something like that feel free 😄. Something merged might be better for the default -t view than what is currently there.

Since I was assuming the DAG to be faulty that is what I spend a couple of hours on yesterday^^ I guess the question was more a: Can it easily be done? (I assume not)

**Comment by tgamblin** [comment]
> `s.traverse()` seems to only follow `link` and `run` edges, even though `get_dependencies` calls it without an explicit `deptype` and `traverse_edges` defaults to `deptype='all'`. Nevertheless `libtool` does not appear in the final view even though there are multiple packages that have a `depends_on('libtool', type='build')` in the requested DAG.

Not quite.  It *does* follow any `deptype` by default (`deptype=all`), but if you are calling `traverse()` on the final spec, you may notice that we strip build-only dependencies from what gets stored in the database and in `spec.yaml`.  We also don't currently include them in the hash.  This is so that Spack doesn't rebuild things every time you upgrade your CMake version.  We'll include finer-grained provenance soon once the concretizer can reuse things better (and it won't matter as much if hashes change).

> Question: If `a` has a `build`-only dependency on `b` and I do `spack install a && spack uninstall b` will spack try to uninstall `a`?

No.

> Since I was assuming the DAG to be faulty that is what I spend a couple of hours on yesterday^^ I guess the question was more a: Can it easily be done? (I assume not)

It's not hard to show merged deptypes for the `cover='nodes'` view.  I did it for you in #8821.

The only problem I see now is that if you show merged dependency types for a newly concretized specs, the build dependencies are going to be a bit misleading.  `b` only means that a dependency is needed to build *something*, not that it is necessarily needed to build its parent in the tree view.  Also, it may only be needed to build a transitive dependency, not the root.  So the presence of `b` for some deptype of package `D` in `spack spec spec -t P` doesn't mean that `D` is needed to build `P`.

The other problem with the node view for the way `tree()` works right now is that the traversal is depth-first, so a node is shown the first time it is visited.  So you may see something deep in the tree that is actually a direct dependency of `P`.  If we're setting out to make things more clear, it *might* be better if the default tree view used a breadth-first traversal, so that each node is visited only once, but a node is shown connected to the topmost package that needs it.  That would make the build dep thing clear... you'd be guaranteed that all the direct build dependencies (of `P`) would be shown at the top-level.  Breadth-first would complicated the output of `tree()` as it wouldn't nest things... so you'd either have to do a 2-pass `tree()` or settle for the breadth-first order.

I suppose you can look at `--cover edges` if you want to know this stuff.  Thoughts?

**Comment by healther** [comment]
> Not quite. It does follow any deptype by default (deptype=all), but if you are calling traverse() on the final spec, you may notice that we strip build-only dependencies from what gets stored in the database and in spec.yaml.

That was the part that I didn't get. Thanks a lot!

> It's not hard to show merged deptypes for the cover='nodes' view. I did it for you in #8821.

After looking at your solution in #8821 I feel quite stupid :D thanks, at least this is (hopefully) less confusing (for me it now is, but I also read all your explanations here).

> I suppose you can look at --cover edges if you want to know this stuff. Thoughts?

The concrete case for us was that we were trying to investigate way a package (I think `libtool` or so) appeared in one view but not in another while both had it in the DAG (and both were shown as build-only). So for this case simply adding all the dependency types is sufficient. But of course we could construct a DAG where `libtool` is a build dependency of one build dependency and a build/link dependency of a different build dependency of our root package, i.e. both paths will be stripped out in the end.

I think the more useful way of thinking about it would be: What does the user most likely want to know? And I'd guess this is in the end the question of: What do we need to keep around? In that sense would it make sense to give `spack spec` the output of only the spec as it will end up in the database? (And having the option to fall back to a kind of `--with-build-dependencies` view?)
For me this would probably be the most useful way, but then I also don't usually look at `spack spec` (I always thought of it as a kind of debug tool for developers rather than an information source for the user)
